java/core/src/java/org/apache/orc/impl/SchemaEvolution.java

```diff
@@ -87,7 +87,8 @@ public SchemaEvolution(TypeDescription fileSchema,
     this.fileSchema = fileSchema;
     this.isAcid = checkAcidSchema(fileSchema);
     this.includeAcidColumns = options.getIncludeAcidColumns();
-    this.readerColumnOffset = isAcid ? acidEventFieldNames.size() : 0;
+    this.readerColumnOffset = (isAcid && readerSchema != null && checkAcidSchema(readerSchema)) || !isAcid
+        ? 0 : acidEventFieldNames.size();
     if (readerSchema != null) {
       if (isAcid) {
```

Comment:
You probably want to cut out here with if (isAcid && !isReaderAcid) rather than change addEventSchema.